### PR TITLE
single thumbnailer instance

### PR DESCRIPTION
**Pull request by dhruv2295**

### CODE REVIEW

File: lib/src/main/java/com/otaliastudios/transcoder/internal/Tracks.kt [hunk 1]
```diff
@@ -11,19 +11,23 @@ import com.otaliastudios.transcoder.source.DataSource
 import com.otaliastudios.transcoder.strategy.TrackStrategy
 
 class Tracks(
-    strategies: TrackMap<TrackStrategy>,
-    sources: DataSources,
-    videoRotation: Int,
-    forceCompression: Boolean
+    val strategies: TrackMap<TrackStrategy>,
```
Comment (sahilbajaj) [inline lib/src/main/java/com/otaliastudios/transcoder/internal/Tracks.kt:14]: I believe, we can avoid making these member properties

File: lib/src/main/java/com/otaliastudios/transcoder/internal/CustomSegments.kt [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+    // it will be time dependent
+    // 1. make segments work for thumbnails as is
+    // 2. inject segments dynamically
+    // 3. seek to segment and destroy previous ones
+    // 4. destroy only if necessary, else reuse
+
+    fun getSegment(id: String): Segment? {
+        return segmentMap.getOrPut(id) {
+            if (currentSegment != null) {
```
Comment (sahilbajaj) [inline lib/src/main/java/com/otaliastudios/transcoder/internal/CustomSegments.kt]: Redundant null check

File: lib/src/main/java/com/otaliastudios/transcoder/internal/CustomSegments.kt [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+    private fun destroySegment(segment: Segment?) {
+        segment?.let {
+            segment.release()
+            val source = sources[segment.type][segment.index]
+            if (tracks.active.has(segment.type)) {
+                source.releaseTrack(segment.type)
+            }
+//            source.deinit()
+            val reversed = segmentMap.entries.associate { (k, v) -> v to k }
```
Comment (sahilbajaj) [inline lib/src/main/java/com/otaliastudios/transcoder/internal/CustomSegments.kt]: Either we should create this when segments get updated
Or
Simply use some utility to query by value

**Review by sahilbajaj — CHANGES_REQUESTED** [review]
Some minor changes

File: lib/src/main/java/com/otaliastudios/transcoder/internal/thumbnails/DefaultThumbnailsEngine.kt [hunk 4]
```diff
@@ -206,58 +203,78 @@ class DefaultThumbnailsEngine(
...
+                request.locate(duration).map { it to request }
+            }.sortedBy { it.first }
+            val index = dataSources[TrackType.VIDEO].indexOfFirst { it.mediaId() == entry.key }
+            stubs.addAll(
+                positions.mapNotNull { (positionUs, request) ->
+                    val localizedUs = timer.localize(TrackType.VIDEO, index, positionUs)
+                    localizedUs?.let { Stub(request, positionUs, localizedUs) }
+                }.toMutableList().reorder(dataSources[TrackType.VIDEO][index])
+            )
+                log.i("Updating pipeline positions for segment Index#$index absoluteUs=${positions.joinToString { it.first.toString() }}, and stubs $stubs")
```
Comment (sahilbajaj) [inline lib/src/main/java/com/otaliastudios/transcoder/internal/thumbnails/DefaultThumbnailsEngine.kt:255]: We should just use a flag. So it's hard to pass these logs to develop

**Review by sahilbajaj — COMMENTED** [review]
Some comments

File: lib/src/main/java/com/otaliastudios/transcoder/internal/CustomSegments.kt (hunk 2 above)
Comment (dhruv2295) [reply lib/src/main/java/com/otaliastudios/transcoder/internal/CustomSegments.kt]: done

File: lib/src/main/java/com/otaliastudios/transcoder/internal/thumbnails/DefaultThumbnailsEngine.kt (hunk 4 above)
Comment (dhruv2295) [reply lib/src/main/java/com/otaliastudios/transcoder/internal/thumbnails/DefaultThumbnailsEngine.kt:255]: done

File: lib/src/main/java/com/otaliastudios/transcoder/internal/CustomSegments.kt [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    fun getSegment(id: String): Segment? {
+        return segmentMap.getOrPut(id) {
+            destroySegment(id)
+            tryCreateSegment(id).also {
+                currentSegment = it
+            }
+        }
+    }
+
+    fun release() = destroySegment()
```
Comment (sahilbajaj) [inline lib/src/main/java/com/otaliastudios/transcoder/internal/CustomSegments.kt:53]: Let's create a new method destroySegments()

**Review by sahilbajaj — APPROVED** [review]

**Pull request merged**